[WIP] Transcoding to UTF-32

NicolasJiaxin opened the pull request:
Fixes #76 
This should add transcoding to UTF-32 from UTF-8 and UTF-16 by the time it is done.

NicolasJiaxin commented:
@lemire For now, I have only implemented for westmere and I did not test it yet so it might now be worth reviewing.

kostya commented on include/simdutf/implementation.h:
```diff
@@ -84,6 +84,18 @@ simdutf_warn_unused size_t convert_utf8_to_utf16(const char * input, size_t leng
+ * @param utf32_buffer  the pointer to buffer that can hold conversion result
+ * @return the number of written char32_t
+ */
+simdutf_warn_unused size_t convert_valid_utf8_to_utf32(const char * input, size_t length, char32_t* utf32_buffer) noexcept;
```
is it possible to add also api function, which validate + convert into temp buffer with fixed size. Allocating full buffer not always good idea.

lemire replied:
You can convert just a substring: provide a pointer with the desired input length.

kostya replied:
With substring it requires return pointer of the last valid char. To continue from that, at the next iteration. Also  substring is by default invalid utf8

lemire replied:
It is probably best if you turn your need into a distinct issue.

lemire commented:
Thanks!

NicolasJiaxin commented:
@lemire I have set up the testing, and right now I am stuck on pure ASCII conversion from UTF-8 to UTF-32. You can run `convert_valid_utf8_to_utf32_tests` to see by yourself. It seems that the `store_ascii_as_utf32` in `simd.h` for haswell (and probably westmere) are wrong. I was not able to correctly use the instructions in that function. Could you look into it when you have time?

lemire commented:
@NicolasJiaxin On my todo.

NicolasJiaxin commented:
I was able to come up with something that seems to work, but there is probably a better to do it than what I did for ASCII transcoding in haswell.

NicolasJiaxin commented:
I was able to come up with something that seems to work, but there is probably a better to do it than what I did for ASCII transcoding in haswell.

lemire commented:
I'll have a look this afternoon. I was actually work on other PRs in this project. :-)

NicolasJiaxin commented:
In my most recent commit, I added a line in the CMake file in `src` to fix an error regarding the modifications I have done in `westmere/simd.h`. Let me know if this is appropriate.

NicolasJiaxin commented:
> In my most recent commit, I added a line in the CMake file in `src` to fix an error regarding the modifications I have done in `westmere/simd.h`. 


There are still some problems in the singleheader build as seen below. The line I added fixed when building normally, but the error is still there in the singleheader build.

lemire commented:
I will get to it!

lemire commented:
I am looking at the single-header issue. Just starting. It is probably a small thing.

lemire commented on src/arm64/arm_convert_utf8_to_utf32.cpp:
```diff
@@ -0,0 +1,130 @@
+  // We first try a few fast paths.
+  if((utf8_end_of_code_point_mask & 0xFFFF) == 0xFFFF) {
+    // We process in chunks of 16 bytes
+    vst1q_u32(reinterpret_cast<uint32_t*>(utf32_output), vmovl_u16(vget_low_u16(vmovl_u8(vget_low_u8 (in)))));
```
It could possibly be more elegant to do the cast to `uint32_t*` once as it might avoid the rather verbose `reinterpret_cast<uint32_t*>(utf32_output)`.

lemire commented on src/arm64/arm_convert_utf8_to_utf32.cpp:
```diff
@@ -0,0 +1,130 @@
+    utf32_output += 16; // We wrote 16 16-bit characters.
+    return 16; // We consumed 16 bytes.
+  }
+  if((utf8_end_of_code_point_mask & 0xFFFF) == 0xaaaa) {
```
Though I also make this mistake, it is probably best not to mix case. If you use 0xaaaa, then use 0xffff.

lemire commented:
We probably want to hook this into our benchmarks. It is simpler than it looks but it might require some additional effort and could be done as a separate PR, possibly... though it would be very nice to check the performance prior to merging.

NicolasJiaxin commented:
> We probably want to hook this into our benchmarks. It is simpler than it looks but it might require some additional effort and could be done as a separate PR, possibly... though it would be very nice to check the performance prior to merging.

Ok, I will try to do that!

NicolasJiaxin commented:
@lemire I did not see prior to starting that there were also a little-endian and big-endian variations for UTF-32. The format I assumed unknowingly is UTF32-LE. Is that ok?

NicolasJiaxin commented:
I have added a benchmark for valid UTF-8 to UTF-32 transcoding only for our implementation for now... The results I got for a 10KB random UTF-8 input are
```
>>> ./build/benchmarks/benchmark --random-utf8 10240 -P convert_valid_utf8_to_utf32
We define the number of bytes to be the number of *input* bytes.
We define a 'char' to be a code point (between 1 and 4 bytes).
===========================
Using ICU version 67.1
testcases: 1
input detected as UTF8
current system detected as haswell
===========================
convert_valid_utf8_to_utf32+fallback, input size: 10244, iterations: 2000
  12.425 ins/byte,   12.425 cycle/byte,    1.196 GB/s (2.7 %), 0.008786 b.misses/byte, 0.000000 c.mis/byte,    3.710 GHz,    4.005 ins/cycle 
  30.886 ins/char,   30.886 cycle/char,    0.481 Gc/s (2.7 %), 0.021839 b.misses/char, 0.000000 c.mis/char,     2.49 byte/char 
convert_valid_utf8_to_utf32+haswell, input size: 10244, iterations: 2000
   5.727 ins/byte,    5.727 cycle/byte,    1.761 GB/s (1.5 %), 0.001074 b.misses/byte, 0.000000 c.mis/byte,    3.725 GHz,    2.707 ins/cycle 
  14.237 ins/char,   14.237 cycle/char,    0.708 Gc/s (1.5 %), 0.002669 b.misses/char, 0.000000 c.mis/char,     2.49 byte/char 
convert_valid_utf8_to_utf32+westmere, input size: 10244, iterations: 2000
   6.094 ins/byte,    6.094 cycle/byte,    1.774 GB/s (1.6 %), 0.000683 b.misses/byte, 0.000000 c.mis/byte,    3.723 GHz,    2.905 ins/cycle 
  15.149 ins/char,   15.149 cycle/char,    0.714 Gc/s (1.6 %), 0.001699 b.misses/char, 0.000000 c.mis/char,     2.49 byte/char 
```

lemire commented:
@NicolasJiaxin 

> I did not see prior to starting that there were also a little-endian and big-endian variations for UTF-32. The format I assumed unknowingly is UTF32-LE. Is that ok?

Correct. We have the same issue with UTF-16 where we only support UTF-16LE. Adding support for UTF-16BE is not very difficult (just flip the bytes). I will add it to our todo.

lemire commented:
See https://github.com/simdutf/simdutf/issues/118

lemire commented:
@NicolasJiaxin On your machine, how does convert_valid_utf8_to_utf32+haswell compare with convert_valid_utf8_to_utf16+haswell ? It should be 'similar'...

Note that you can test with actual files... Grab this:

https://github.com/lemire/unicode_lipsum

Then just pass a file name (`-F myfile.utf8.txt`).

lemire commented:
@NicolasJiaxin Try the following to fix your amalgamation issue...

```
diff --git a/src/simdutf.cpp b/src/simdutf.cpp
index 45f8af3..de4c7a9 100644
--- a/src/simdutf.cpp
+++ b/src/simdutf.cpp
@@ -11,6 +11,7 @@
 #include "scalar/utf16_to_utf8/valid_utf16_to_utf8.h"
 #include "scalar/utf16_to_utf8/utf16_to_utf8.h"
 #include "scalar/utf8_to_utf16/valid_utf8_to_utf16.h"
+#include "scalar/utf8_to_utf32/valid_utf8_to_utf32.h"
 #include "scalar/utf8_to_utf16/utf8_to_utf16.h"
 #include "scalar/utf8.h"
 #include "scalar/utf16.h"
```

lemire commented on src/westmere/sse_convert_utf8_to_utf32.cpp:
```diff
@@ -0,0 +1,124 @@
+      utf8_end_of_code_point_mask & 0xfff;
+  if(((utf8_end_of_code_point_mask & 0xffff) == 0xffff)) {
+    // We process the data in chunks of 16 bytes.
+    _mm256_storeu_si256(reinterpret_cast<__m256i *>(utf32_output), _mm256_cvtepu8_epi32(in));
```
So under the westmere kernel, you are using avx2 intrinsics. You do not (yet) catch the issue because you are compiling with `-march=native` but if you lift that flag, you will see issues.

NicolasJiaxin replied:
I see... and yes, indeed I started to have problems when I started to work with westmere. This was the "solution" I found online.

NicolasJiaxin commented:
> @NicolasJiaxin Try the following to fix your amalgamation issue...

I applied the patch and I think it helped, but I still have the issue with avx2 intrinsics in the singleheader version, and removing the flag will also cause an error when building normally.

lemire commented:
I will have a look again. :-)

NicolasJiaxin commented:
I have added UTF-8 to UTF-32 transcoding with validation as well as the tests (convert_utf8_to_utf32_tests). I have also added a function utf32_length_from_utf8, analogous to utf16_length_from_utf8. It is exactly the same as count_code_points for utf8. I copied the code from that function rather than making a function call so that utf32_length_from_utf8 does not depend on that function although I don't think it could become a problem anyways.

lemire commented:
@NicolasJiaxin 

The following patch should allow you to build a proper westmere kernel:

```diff
diff --git a/src/CMakeLists.txt b/src/CMakeLists.txt
index c40b245..4322f3f 100644
--- a/src/CMakeLists.txt
+++ b/src/CMakeLists.txt
@@ -1,4 +1,3 @@
-set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -march=native")
 add_library(simdutf-include-source INTERFACE)
 target_include_directories(simdutf-include-source INTERFACE $<BUILD_INTERFACE:${CMAKE_CURRENT_SOURCE_DIR}>)
 add_library(simdutf-source INTERFACE)
diff --git a/src/simdutf/westmere/simd.h b/src/simdutf/westmere/simd.h
index 37cf3cd..5e88935 100644
--- a/src/simdutf/westmere/simd.h
+++ b/src/simdutf/westmere/simd.h
@@ -23,8 +23,10 @@ namespace simd {
       _mm_storeu_si128(reinterpret_cast<__m128i *>(p+8), _mm_cvtepu8_epi16(_mm_srli_si128(*this,8)));
     }
     simdutf_really_inline void store_ascii_as_utf32(char32_t * p) const {
-      _mm256_storeu_si256(reinterpret_cast<__m256i *>(p), _mm256_cvtepu8_epi32(*this));
-      _mm256_storeu_si256(reinterpret_cast<__m256i *>(p+8), _mm256_cvtepu8_epi32(_mm_srli_si128(*this,8)));
+      _mm_storeu_si128(reinterpret_cast<__m128i *>(p), _mm_cvtepu8_epi32(*this));
+      _mm_storeu_si128(reinterpret_cast<__m128i *>(p+4), _mm_cvtepu8_epi32(_mm_srli_si128(*this,4)));
+      _mm_storeu_si128(reinterpret_cast<__m128i *>(p+8), _mm_cvtepu8_epi32(_mm_srli_si128(*this,8)));
+      _mm_storeu_si128(reinterpret_cast<__m128i *>(p+12), _mm_cvtepu8_epi32(_mm_srli_si128(*this,12)));
     }
     // Bit operations
     simdutf_really_inline Child operator|(const Child other) const { return _mm_or_si128(*this, other); }
diff --git a/src/westmere/sse_convert_utf8_to_utf32.cpp b/src/westmere/sse_convert_utf8_to_utf32.cpp
index 3b1878c..836dc2d 100644
--- a/src/westmere/sse_convert_utf8_to_utf32.cpp
+++ b/src/westmere/sse_convert_utf8_to_utf32.cpp
@@ -23,9 +23,11 @@ size_t convert_masked_utf8_to_utf32(const char *input,
       utf8_end_of_code_point_mask & 0xfff;
   if(((utf8_end_of_code_point_mask & 0xffff) == 0xffff)) {
     // We process the data in chunks of 16 bytes.
-    _mm256_storeu_si256(reinterpret_cast<__m256i *>(utf32_output), _mm256_cvtepu8_epi32(in));
-    _mm256_storeu_si256(reinterpret_cast<__m256i *>(utf32_output+8), _mm256_cvtepu8_epi32(_mm_srli_si128(in,8)));
-    utf32_output += 16; // We wrote 16 16-bit characters.
+    _mm_storeu_si128(reinterpret_cast<__m128i *>(utf32_output), _mm_cvtepu8_epi32(in));
+    _mm_storeu_si128(reinterpret_cast<__m128i *>(utf32_output+4), _mm_cvtepu8_epi32(_mm_srli_si128(in,4)));
+    _mm_storeu_si128(reinterpret_cast<__m128i *>(utf32_output+8), _mm_cvtepu8_epi32(_mm_srli_si128(in,8)));
+    _mm_storeu_si128(reinterpret_cast<__m128i *>(utf32_output+12), _mm_cvtepu8_epi32(_mm_srli_si128(in,12)));
+    utf32_output += 16; // We wrote 16 32-bit characters.
     return 16; // We consumed 16 bytes.
   }
   if(((utf8_end_of_code_point_mask & 0xffff) == 0xaaaa)) {
@@ -36,8 +38,9 @@ size_t convert_masked_utf8_to_utf32(const char *input,
     const __m128i ascii = _mm_and_si128(perm, _mm_set1_epi16(0x7f));
     const __m128i highbyte = _mm_and_si128(perm, _mm_set1_epi16(0x1f00));
     const __m128i composed = _mm_or_si128(ascii, _mm_srli_epi16(highbyte, 2));
-    _mm256_storeu_si256((__m256i *)utf32_output, _mm256_cvtepu16_epi32(composed));
-    utf32_output += 8; // We wrote 16 bytes, 8 code points.
+    _mm_storeu_si128(reinterpret_cast<__m128i *>(utf32_output), _mm_cvtepu8_epi32(composed));
+    _mm_storeu_si128(reinterpret_cast<__m128i *>(utf32_output+4), _mm_cvtepu8_epi32(_mm_srli_si128(composed,4)));
+    utf32_output += 8; // We wrote 32 bytes, 8 code points.
     return 16;
   }
   if(input_utf8_end_of_code_point_mask == 0x924) {
@@ -77,7 +80,8 @@ size_t convert_masked_utf8_to_utf32(const char *input,
     const __m128i ascii = _mm_and_si128(perm, _mm_set1_epi16(0x7f));
     const __m128i highbyte = _mm_and_si128(perm, _mm_set1_epi16(0x1f00));
     const __m128i composed = _mm_or_si128(ascii, _mm_srli_epi16(highbyte, 2));
-    _mm256_storeu_si256((__m256i *)utf32_output, _mm256_cvtepu16_epi32(composed));
+    _mm_storeu_si128(reinterpret_cast<__m128i *>(utf32_output), _mm_cvtepu8_epi32(composed));
+    _mm_storeu_si128(reinterpret_cast<__m128i *>(utf32_output+4), _mm_cvtepu8_epi32(_mm_srli_si128(composed,4)));
     utf32_output += 6; // We wrote 12 bytes, 6 code points.
   } else if (idx < 145) {
     // FOUR (4) input code-words
```

NicolasJiaxin commented:
> @NicolasJiaxin
> 
> The following patch should allow you to build a proper westmere kernel:

Yes, thanks!! It seems to fix the issue with westmere and the singleheader build. However, westmere no longer passes the test I implemented convert_valid_utf8_to_utf32 and convert_utf8_to_utf32. I can look into it.

lemire commented:
Ah. Well, I just said *compile* not *run*. :-)

NicolasJiaxin commented:
> Ah. Well, I just said _compile_ not _run_. :-)

Cannot deny that hahahaha. I think it should be fixed now. It was simply to use `_mm_cvtepu16_epi32` instead of `_mm_cvtepu8_epi32` since composed should be UTF-16 words and `_mm_srli_si128(composed,8)` instead of `_mm_srli_si128(composed,4)`.

NicolasJiaxin commented:
@lemire I have added all competitors that offer UTF-8 to UTF-32 transcoding to the benchmark, except CppCon2018. I am getting a similar error than the one we had with westmere build. I also did not include ICU in the benchmark although it is technically possible to transcode from UTF-8 to UTF-32 by using two conversions (UTF-8->UTF-16->UTF32), but I can if you think that is better. Here is a patch of my attempt to include CppCon2018:

```
diff --git a/benchmarks/src/benchmark.cpp b/benchmarks/src/benchmark.cpp
index c2b1e33..474a8f5 100644
--- a/benchmarks/src/benchmark.cpp
+++ b/benchmarks/src/benchmark.cpp
@@ -154,6 +154,11 @@ Benchmark::Benchmark(std::vector<input::Testcase>&& testcases)
         known_procedures.insert(name);
         expected_input_encoding.insert(std::make_pair(name,std::set<simdutf::encoding_type>({simdutf::encoding_type::UTF8})));
     }
+    {
+        std::string name = "convert_utf8_to_utf32+cppcon2018";
+        known_procedures.insert(name);
+        expected_input_encoding.insert(std::make_pair(name,std::set<simdutf::encoding_type>({simdutf::encoding_type::UTF8})));
+    }
 #endif
     {
         std::string name = "convert_utf8_to_utf16+hoehrmann";
@@ -231,6 +236,8 @@ void Benchmark::run(const std::string& procedure_name, size_t iterations) {
         // this is a special case
         if(name == "convert_utf8_to_utf16") {
           run_convert_utf8_to_utf16_cppcon2018(iterations);
+        } else if(name == "convert_utf8_to_utf32") {
+          run_convert_utf8_to_utf16_cppcon2018(iterations);
         } else {
           std::cerr << "unrecognized:" << procedure_name << "\n";
         }
@@ -757,6 +764,28 @@ void Benchmark::run_convert_utf8_to_utf16_cppcon2018(size_t iterations) {
     size_t char_count = active_implementation->count_utf8(reinterpret_cast<const char*>(data), size);
     print_summary(result, size, char_count);
 }
+/**
+ * Bob Steagall, CppCon2018
+ * https://github.com/BobSteagall/CppCon2018/
+ *
+ * Fast Conversion From UTF-8 with C++, DFAs, and SSE Intrinsics
+ * https://www.youtube.com/watch?v=5FQ87-Ecb-A
+ */
+void Benchmark::run_convert_utf8_to_utf32_cppcon2018(size_t iterations) {
+    using char8_t   = unsigned char;
+    const char8_t*  data = reinterpret_cast<const char8_t*>(input_data.data());
+    const size_t size = input_data.size();
+    std::unique_ptr<char32_t[]> output_buffer{new char32_t[size]};
+    volatile size_t sink{0};
+    auto proc = [data, size, &output_buffer, &sink]() {
+      sink = uu::UtfUtils::SseConvert(data, data + size, output_buffer.get());
+    };
+    count_events(proc, iterations); // warming up!
+    const auto result = count_events(proc, iterations);
+    if((sink == 0) && (size != 0) && (iterations > 0)) { std::cerr << "The output is zero which might indicate a misconfiguration.\n"; }
+    size_t char_count = active_implementation->count_utf8(reinterpret_cast<const char*>(data), size);
+    print_summary(result, size, char_count);
+}
 /**
  * Cameron, Robert D, A case study in SIMD text processing with parallel bit streams: UTF-8 to UTF-16 transcoding,
  * Proceedings of the 13th ACM SIGPLAN Symposium on Principles and practice of parallel programming, 91--98.
diff --git a/benchmarks/src/benchmark.h b/benchmarks/src/benchmark.h
index e22402c..3595535 100644
--- a/benchmarks/src/benchmark.h
+++ b/benchmarks/src/benchmark.h
@@ -92,6 +92,7 @@ namespace simdutf::benchmarks {
         * https://www.youtube.com/watch?v=5FQ87-Ecb-A
         */
         void run_convert_utf8_to_utf16_cppcon2018(size_t iterations);
+        void run_convert_utf8_to_utf32_cppcon2018(size_t iterations);
         /**
          * benchmarks/competition/u8u16 contains an open source version of u8u16, referenced in
          * Cameron, Robert D, A case study in SIMD text processing with parallel bit streams: UTF-8 to UTF-16 transcoding,

```

NicolasJiaxin commented:
Also, do you want me to make a table comparing all the competitors on unicode_lipsum data?

lemire commented:
> I also did not include ICU in the benchmark although it is technically possible to transcode from UTF-8 to UTF-32 by using two conversions (UTF-8->UTF-16->UTF32), but I can if you think that is better.

It does not sound very enticing, now does it? It feels like it would not be reasonable as a competitor.

lemire commented:
> Also, do you want me to make a table comparing all the competitors on unicode_lipsum data?

Though it is fun, it is maybe too much work. I recommend picking a few... like latin/english, arabic, japanese...

lemire commented:
@NicolasJiaxin 

Here:

```diff
+        } else if(name == "convert_utf8_to_utf32") {
+          run_convert_utf8_to_utf16_cppcon2018(iterations);
```

You probably want to write `run_convert_utf8_to_utf32_cppcon2018`.

Second... try the following...

```diff
--- a/benchmarks/competition/CppCon2018/utf_utils.h
+++ b/benchmarks/competition/CppCon2018/utf_utils.h
@@ -368,7 +368,7 @@ UtfUtils::FastConvert(char8_t const* pSrc, char8_t const* pSrcEnd, char32_t* pDs
 ///     indicate an error was encountered.
 //--------------------------------------------------------------------------------------------------
 //
-KEWB_FORCE_INLINE ptrdiff_t
+ptrdiff_t
 UtfUtils::SseConvert(char8_t const* pSrc, char8_t const* pSrcEnd, char32_t* pDst) noexcept
 {
```

The problem is that CppCon2018 wants to forcefully inline the conversion function but we *do not want* that.

:-)

NicolasJiaxin commented:
I wrote a script so it was not too bad. Here are the results for all implementations in our benchmark for transcoding various UTF-8 texts to UTF-32 (5000 iterations, data is from from [unicode-lipsum](https://github.com/lemire/unicode_lipsum)).

| |CppCon2018 (GB/s)|fallback (GB/s)|haswell (GB/s)|hoehrmann (GB/s)|llvm (GB/s)|utf8lut (GB/s)|westmere (GB/s)|fallback(da) (GB/s)|haswell(da) (GB/s)|westmere(da) (GB/s)|
|---------------------------|-----|-----|-----|-----|-----|-----|-----|-----|-----|-----|
|Arabic-Lipsum.utf8.txt     |0.672|0.573|2.255|0.376|0.266|1.770|1.810|0.503|2.130|1.697|
|Chinese-Lipsum.utf8.txt    |0.913|0.686|2.773|0.458|0.572|2.303|2.270|0.589|2.599|2.104|
|Emoji-Lipsum.utf8.txt      |0.982|0.738|2.257|0.461|0.590|0.681|1.581|0.627|2.145|1.493|
|Hebrew-Lipsum.utf8.txt     |0.677|0.584|2.227|0.370|0.261|1.861|1.774|0.512|2.110|1.670|
|Hindi-Lipsum.utf8.txt      |0.719|0.558|1.617|0.373|0.397|2.335|1.489|0.492|1.554|1.412|
|Japanese-Lipsum.utf8.txt   |0.875|0.663|2.668|0.439|0.530|2.275|2.228|0.571|2.489|2.067|
|Korean-Lipsum.utf8.txt     |0.840|0.698|1.907|0.380|0.324|2.035|1.747|0.597|1.821|1.646|
|Latin-Lipsum.utf8.txt      |4.599|3.087|9.330|0.461|0.246|1.258|8.166|1.786|7.928|6.380|
|Russian-Lipsum.utf8.txt    |0.593|0.445|2.224|0.330|0.265|1.827|1.816|0.402|2.117|1.705|

da stands for dynamic allocation in the last three columns.
Sample command: ` ./build/benchmarks/benchmark -F FILE_PATH -P convert_utf8_to_utf32 -I 5000`

NicolasJiaxin commented:
> Second... try the following...
@lemire  Thanks, it worked!

lemire commented:
I like these numbers. And wait till we throw AVX-512 at it!!!

NicolasJiaxin commented:
@lemire How can I run the tests on an ARM machine?

lemire commented:
@NicolasJiaxin Check your email.

:-)

lemire commented:
It may be beneficial to sync with the main project.

NicolasJiaxin commented:
These are the results for UTF-16 to UTF-32 conversions of a few implementations. Again, benchmarking using [unicode-lipsum](https://github.com/lemire/unicode_lipsum).

| | fallback | haswell | llvm | westmere | fallback(da)| haswell(da) | westmere(da) |
|----------------------------|------|-------|------|-------|------|--------|------|
|Arabic-Lipsum.utf16.txt    |2.938|17.887|1.475|13.752|2.251|10.756|7.969|
|Chinese-Lipsum.utf16.txt   |2.936|19.535|1.474|14.032|2.242|11.915|7.857|
|Emoji-Lipsum.utf16.txt     |2.601|0.641|2.456|2.499|2.167|0.625|2.191|
|Hebrew-Lipsum.utf16.txt    |2.939|17.668|1.475|13.680|2.249|10.591|7.876|
|Hindi-Lipsum.utf16.txt     |2.936|19.492|1.475|13.940|2.247|11.151|7.807|
|Japanese-Lipsum.utf16.txt  |2.937|20.151|1.474|14.223|2.241|11.953|7.815|
|Korean-Lipsum.utf16.txt    |2.936|17.806|1.474|13.593|2.247|10.614|7.687|
|Latin-Lipsum.utf16.txt     |2.940|14.996|1.476|12.776|2.247|9.525|7.470|
|Russian-Lipsum.utf16.txt   |2.941|15.955|1.475|13.151|2.249|10.112|7.651|

For some reasons, haswell performs horribly on Emoji-Lipsum.utf16.txt (consists of surrogate pairs UTF-16).

lemire commented:
> For some reason, haswell performs horribly on Emoji-Lipsum.utf16.txt (consists only of surrogate pairs UTF-16, and is the only file in the list that contains such characters).

That's intriguing. Let us investigate.

NicolasJiaxin commented:
@lemire I don't really understand where the decrease is coming from. The code for that case is basically the same as westmere, except that we are working with 256-bit vectors instead of 128. However, I have ran the test for UTF-16 to UTF-8 with the same file  (Emoji-Lipsum.utf16.txt) and haswell performs similarly to westmere and the fallback.

lemire commented:
Yes. I will investigate. I was investigating something else earlier today. I will soon move to this issue you raise. Thanks.

lemire commented on include/simdutf/implementation.h:
```diff
@@ -97,6 +122,19 @@ simdutf_warn_unused size_t convert_valid_utf8_to_utf16(const char * input, size_
+ * @param length        the length of the string in bytes
+ * @return the number of char32_t words required to encode the UTF-8 string as UTF-32LE
+ */
+simdutf_warn_unused size_t utf32_length_from_utf8(const char * input, size_t length) noexcept;
```
Is this needed as part of the API? That's just count_utf8 isn't it?

lemire replied:
(I am not asking you to remove the function, I am just curious.)

NicolasJiaxin replied:
Yes, it is exactly the same. I included it for uniformity, but we can remove it of course...

lemire replied:
If we keep it, then we should explain in the comments that it is equivalent to... 

See what I mean? Otherwise someone might get confused and think that these are different functions.

NicolasJiaxin replied:
Ahh yes, that would be clearer indeed.

lemire commented on include/simdutf/implementation.h:
```diff
@@ -137,6 +204,17 @@ simdutf_warn_unused size_t convert_valid_utf16_to_utf8(const char16_t * input, s
+ * @param length        the length of the string in 2-byte words (char16_t)
+ * @return the number of bytes required to encode the UTF-16LE string as UTF-32LE
+ */
+simdutf_warn_unused size_t utf32_length_from_utf16(const char16_t * input, size_t length) noexcept;
```
Same question... isn't this just `count_utf16`?

NicolasJiaxin replied:
Yes, for same reason as above

lemire commented:
@NicolasJiaxin  Would you update the README file with the new functionality?

lemire commented:
This PR looks good to me. Let me investigate the performance issue.

lemire commented:
The UTF-8 to UTF-32 over emoji looks good... (I am aware that it is not the issue you reported.)

```
$ ./build/benchmarks/benchmark -F unicode_lipsum/lipsum/Emoji-Lipsum.utf8.txt -P convert_utf8_to_utf32
We define the number of bytes to be the number of *input* bytes.
We define a 'char' to be a code point (between 1 and 4 bytes).
===========================
Using ICU version 66.1
testcases: 1
input detected as UTF8
current system detected as haswell
===========================
convert_utf8_to_utf32+cppcon2018, input size: 65542, iterations: 2000, dataset: unicode_lipsum/lipsum/Emoji-Lipsum.utf8.txt
  17.004 ins/byte,   17.004 cycle/byte,    0.776 GB/s (0.2 %), 0.000076 b.misses/byte, 0.000000 c.mis/byte,    3.695 GHz,    3.572 ins/cycle 
  68.012 ins/char,   68.012 cycle/char,    0.194 Gc/s (0.2 %), 0.000305 b.misses/char, 0.000000 c.mis/char,     4.00 byte/char 
convert_utf8_to_utf32+fallback, input size: 65542, iterations: 2000, dataset: unicode_lipsum/lipsum/Emoji-Lipsum.utf8.txt
  15.004 ins/byte,   15.004 cycle/byte,    0.867 GB/s (0.1 %), 0.000107 b.misses/byte, 0.000000 c.mis/byte,    3.695 GHz,    3.520 ins/cycle 
  60.014 ins/char,   60.014 cycle/char,    0.217 Gc/s (0.1 %), 0.000427 b.misses/char, 0.000000 c.mis/char,     4.00 byte/char 
convert_utf8_to_utf32+haswell, input size: 65542, iterations: 2000, dataset: unicode_lipsum/lipsum/Emoji-Lipsum.utf8.txt
   4.975 ins/byte,    4.975 cycle/byte,    2.322 GB/s (0.6 %), 0.000061 b.misses/byte, 0.000000 c.mis/byte,    3.699 GHz,    3.123 ins/cycle 
  19.899 ins/char,   19.899 cycle/char,    0.580 Gc/s (0.6 %), 0.000244 b.misses/char, 0.000000 c.mis/char,     4.00 byte/char 
convert_utf8_to_utf32+hoehrmann, input size: 65542, iterations: 2000, dataset: unicode_lipsum/lipsum/Emoji-Lipsum.utf8.txt
  16.754 ins/byte,   16.754 cycle/byte,    0.461 GB/s (0.1 %), 0.000137 b.misses/byte, 0.000015 c.mis/byte,    3.694 GHz,    2.091 ins/cycle 
  67.013 ins/char,   67.013 cycle/char,    0.115 Gc/s (0.1 %), 0.000549 b.misses/char, 0.000061 c.mis/char,     4.00 byte/char 
convert_utf8_to_utf32+llvm, input size: 65542, iterations: 2000, dataset: unicode_lipsum/lipsum/Emoji-Lipsum.utf8.txt
  22.505 ins/byte,   22.505 cycle/byte,    0.557 GB/s (2.8 %), 0.000107 b.misses/byte, 0.000000 c.mis/byte,    3.694 GHz,    3.392 ins/cycle 
  90.017 ins/char,   90.017 cycle/char,    0.139 Gc/s (2.8 %), 0.000427 b.misses/char, 0.000000 c.mis/char,     4.00 byte/char 
convert_utf8_to_utf32+utf8lut, input size: 65542, iterations: 2000, dataset: unicode_lipsum/lipsum/Emoji-Lipsum.utf8.txt
  13.952 ins/byte,   13.952 cycle/byte,    0.679 GB/s (0.4 %), 0.000076 b.misses/byte, 0.000000 c.mis/byte,    3.693 GHz,    2.564 ins/cycle 
  55.806 ins/char,   55.806 cycle/char,    0.170 Gc/s (0.4 %), 0.000305 b.misses/char, 0.000000 c.mis/char,     4.00 byte/char 
convert_utf8_to_utf32+westmere, input size: 65542, iterations: 2000, dataset: unicode_lipsum/lipsum/Emoji-Lipsum.utf8.txt
   7.174 ins/byte,    7.174 cycle/byte,    1.725 GB/s (8.3 %), 0.000107 b.misses/byte, 0.001053 c.mis/byte,    3.697 GHz,    3.347 ins/cycle 
  28.695 ins/char,   28.695 cycle/char,    0.431 Gc/s (8.3 %), 0.000427 b.misses/char, 0.004211 c.mis/char,     4.00 byte/char 
```

lemire commented:
@NicolasJiaxin What do you think of this patch?
```diff
index 43caf8b..51eb976 100644
--- a/src/haswell/avx2_convert_utf16_to_utf32.cpp
+++ b/src/haswell/avx2_convert_utf16_to_utf32.cpp
@@ -56,9 +56,8 @@ std::pair<const char16_t*, char32_t*> avx2_convert_utf16_to_utf32(const char16_t
   const char16_t* end = buf + len;
   const __m256i v_f800 = _mm256_set1_epi16((int16_t)0xf800);
   const __m256i v_d800 = _mm256_set1_epi16((int16_t)0xd800);
-  const size_t safety_margin = 11; // to avoid overruns, see issue https://github.com/simdutf/simdutf/issues/92
 
-  while (buf + 16 + safety_margin <= end) {
+  while (buf + 16 <= end) {
     __m256i in = _mm256_loadu_si256((__m256i*)buf);
     
     // 1. Check if there are any surrogate word in the input chunk.
```

lemire commented:
I am not reliably reproducing your performance issue... It is true that we are "relatively slow" on the `Emoji-Lipsum` dataset, but that's a trade-off... we simply do not consider that it is a likely scenario and we are happy to do away with high performance in that case. (Hopefully, with AVX-512, we will turn this around.)

Could you re-run your benchmark to see what you get?

```
$ ./build/benchmarks/benchmark -F unicode_lipsum/lipsum/Emoji-Lipsum.utf16.txt -P convert_utf16_to_utf32  -I 20000
We define the number of bytes to be the number of *input* bytes.
We define a 'char' to be a code point (between 1 and 4 bytes).
===========================
Using ICU version 66.1
testcases: 1
input detected as UTF16 little-endian
current system detected as haswell
===========================
convert_utf16_to_utf32+fallback, input size: 65542, iterations: 20000, dataset: unicode_lipsum/lipsum/Emoji-Lipsum.utf16.txt
   6.504 ins/byte,    6.504 cycle/byte,    0.568 GB/s (0.1 %), 0.000061 b.misses/byte, 0.000000 c.mis/byte,    3.693 GHz,    1.000 ins/cycle 
  26.013 ins/char,   26.013 cycle/char,    0.142 Gc/s (0.1 %), 0.000244 b.misses/char, 0.000000 c.mis/char,     4.00 byte/char 
convert_utf16_to_utf32+haswell, input size: 65542, iterations: 20000, dataset: unicode_lipsum/lipsum/Emoji-Lipsum.utf16.txt
   7.160 ins/byte,    7.160 cycle/byte,    0.635 GB/s (0.1 %), 0.000076 b.misses/byte, 0.000000 c.mis/byte,    3.694 GHz,    1.231 ins/cycle 
  28.640 ins/char,   28.640 cycle/char,    0.159 Gc/s (0.1 %), 0.000305 b.misses/char, 0.000000 c.mis/char,     4.00 byte/char 
convert_utf16_to_utf32+llvm, input size: 65542, iterations: 20000, dataset: unicode_lipsum/lipsum/Emoji-Lipsum.utf16.txt
   5.254 ins/byte,    5.254 cycle/byte,    2.105 GB/s (0.1 %), 0.000092 b.misses/byte, 0.000000 c.mis/byte,    3.699 GHz,    2.990 ins/cycle 
  21.016 ins/char,   21.016 cycle/char,    0.526 Gc/s (0.1 %), 0.000366 b.misses/char, 0.000000 c.mis/char,     4.00 byte/char 
convert_utf16_to_utf32+westmere, input size: 65542, iterations: 20000, dataset: unicode_lipsum/lipsum/Emoji-Lipsum.utf16.txt
   7.191 ins/byte,    7.191 cycle/byte,    0.635 GB/s (0.1 %), 0.000046 b.misses/byte, 0.000000 c.mis/byte,    3.693 GHz,    1.236 ins/cycle 
  28.765 ins/char,   28.765 cycle/char,    0.159 Gc/s (0.1 %), 0.000183 b.misses/char, 0.000000 c.mis/char,     4.00 byte/char 
```

NicolasJiaxin commented:
I think I am also having trouble reliably reproducing the issue, but the last two times I have build and run the benchmark I got something like this...

```
$ ./build/benchmarks/benchmark -F ../unicode_lipsum/lipsum/Emoji-Lipsum.utf16.txt -P convert_utf16_to_utf32  -I 20000
We define the number of bytes to be the number of *input* bytes.
We define a 'char' to be a code point (between 1 and 4 bytes).
===========================
Using ICU version 67.1
testcases: 1
input detected as UTF16 little-endian
current system detected as haswell
===========================
convert_utf16_to_utf32+fallback, input size: 65542, iterations: 20000, dataset: ../unicode_lipsum/lipsum/Emoji-Lipsum.utf16.txt
   6.504 ins/byte,    6.504 cycle/byte,    0.590 GB/s (0.1 %), 0.000076 b.misses/byte, 0.000000 c.mis/byte,    3.694 GHz,    1.040 ins/cycle 
  26.014 ins/char,   26.014 cycle/char,    0.148 Gc/s (0.1 %), 0.000305 b.misses/char, 0.000000 c.mis/char,     4.00 byte/char 
convert_utf16_to_utf32+haswell, input size: 65542, iterations: 20000, dataset: ../unicode_lipsum/lipsum/Emoji-Lipsum.utf16.txt
   6.598 ins/byte,    6.598 cycle/byte,    2.364 GB/s (270.4 %), 0.000153 b.misses/byte, 0.002014 c.mis/byte,    3.705 GHz,    4.209 ins/cycle 
  26.391 ins/char,   26.391 cycle/char,    0.591 Gc/s (270.4 %), 0.000610 b.misses/char, 0.008056 c.mis/char,     4.00 byte/char 
WARNING: Measurements are noisy, try increasing iteration count (-I).
convert_utf16_to_utf32+llvm, input size: 65542, iterations: 20000, dataset: ../unicode_lipsum/lipsum/Emoji-Lipsum.utf16.txt
   5.254 ins/byte,    5.254 cycle/byte,    2.456 GB/s (0.1 %), 0.000076 b.misses/byte, 0.000000 c.mis/byte,    3.700 GHz,    3.488 ins/cycle 
  21.016 ins/char,   21.016 cycle/char,    0.614 Gc/s (0.1 %), 0.000305 b.misses/char, 0.000000 c.mis/char,     4.00 byte/char 
convert_utf16_to_utf32+westmere, input size: 65542, iterations: 20000, dataset: ../unicode_lipsum/lipsum/Emoji-Lipsum.utf16.txt
   6.629 ins/byte,    6.629 cycle/byte,    2.499 GB/s (0.2 %), 0.000076 b.misses/byte, 0.000000 c.mis/byte,    3.699 GHz,    4.478 ins/cycle 
  26.515 ins/char,   26.515 cycle/char,    0.625 Gc/s (0.2 %), 0.000305 b.misses/char, 0.000000 c.mis/char,     4.00 byte/char 
convert_utf16_to_utf32_with_dynamic_allocation+fallback, input size: 65542, iterations: 20000, dataset: ../unicode_lipsum/lipsum/Emoji-Lipsum.utf16.txt
   7.760 ins/byte,    7.760 cycle/byte,    0.556 GB/s (0.1 %), 0.000122 b.misses/byte, 0.000000 c.mis/byte,    3.693 GHz,    1.169 ins/cycle 
  31.038 ins/char,   31.038 cycle/char,    0.139 Gc/s (0.1 %), 0.000488 b.misses/char, 0.000000 c.mis/char,     4.00 byte/char 
convert_utf16_to_utf32_with_dynamic_allocation+haswell, input size: 65542, iterations: 20000, dataset: ../unicode_lipsum/lipsum/Emoji-Lipsum.utf16.txt
   6.917 ins/byte,    6.917 cycle/byte,    2.013 GB/s (223.4 %), 0.000168 b.misses/byte, 0.001190 c.mis/byte,    3.696 GHz,    3.767 ins/cycle 
  27.666 ins/char,   27.666 cycle/char,    0.503 Gc/s (223.4 %), 0.000671 b.misses/char, 0.004760 c.mis/char,     4.00 byte/char 
WARNING: Measurements are noisy, try increasing iteration count (-I).
convert_utf16_to_utf32_with_dynamic_allocation+westmere, input size: 65542, iterations: 20000, dataset: ../unicode_lipsum/lipsum/Emoji-Lipsum.utf16.txt
   7.323 ins/byte,    7.323 cycle/byte,    2.200 GB/s (0.3 %), 0.000137 b.misses/byte, 0.000000 c.mis/byte,    3.697 GHz,    4.358 ins/cycle 
  29.290 ins/char,   29.290 cycle/char,    0.550 Gc/s (0.3 %), 0.000549 b.misses/char, 0.000000 c.mis/char,     4.00 byte/char 
```

Now, it is fallback that is behind and westmere and haswell are fine.

lemire commented:
@NicolasJiaxin Right. So when you look at the numbers, you see that the number of instructions is pretty much the same. What varies widely is the number of instructions retired per cycle. I will try it on an AMD processor to see if we can reproduce the issue.

I am pretty sure it has nothing to do with your code, per se.

lemire commented:
@NicolasJiaxin Please consider the patch I proposed above.

NicolasJiaxin commented:
Yes, I forgot to say. These numbers are with the patch applied.

lemire commented:
On an AMD Rome processor, I do not encounter an issue... so I think we are good. You should have access to the AMD box.

```
$ ./build/benchmarks/benchmark -F unicode_lipsum/lipsum/Emoji-Lipsum.utf16.txt -P convert_utf16_to_utf32
We define the number of bytes to be the number of *input* bytes.
We define a 'char' to be a code point (between 1 and 4 bytes).
===========================
Using ICU version 67.1
testcases: 1
input detected as UTF16 little-endian
current system detected as haswell
===========================
convert_utf16_to_utf32+fallback, input size: 65542, iterations: 2000, dataset: unicode_lipsum/lipsum/Emoji-Lipsum.utf16.txt
   6.503 ins/byte,    6.503 cycle/byte,    2.580 GB/s (1.2 %), 0.000061 b.misses/byte, 0.000000 c.mis/byte,    3.402 GHz,    4.932 ins/cycle
  26.013 ins/char,   26.013 cycle/char,    0.645 Gc/s (1.2 %), 0.000244 b.misses/char, 0.000000 c.mis/char,     4.00 byte/char
convert_utf16_to_utf32+haswell, input size: 65542, iterations: 2000, dataset: unicode_lipsum/lipsum/Emoji-Lipsum.utf16.txt
   6.598 ins/byte,    6.598 cycle/byte,    2.463 GB/s (2.6 %), 0.000076 b.misses/byte, 0.000000 c.mis/byte,    3.402 GHz,    4.776 ins/cycle
  26.390 ins/char,   26.390 cycle/char,    0.616 Gc/s (2.6 %), 0.000305 b.misses/char, 0.000000 c.mis/char,     4.00 byte/char
convert_utf16_to_utf32+llvm, input size: 65542, iterations: 2000, dataset: unicode_lipsum/lipsum/Emoji-Lipsum.utf16.txt
   5.254 ins/byte,    5.254 cycle/byte,    3.386 GB/s (0.3 %), 0.000061 b.misses/byte, 0.000000 c.mis/byte,    3.404 GHz,    5.226 ins/cycle
  21.015 ins/char,   21.015 cycle/char,    0.847 Gc/s (0.3 %), 0.000244 b.misses/char, 0.000000 c.mis/char,     4.00 byte/char
convert_utf16_to_utf32+westmere, input size: 65542, iterations: 2000, dataset: unicode_lipsum/lipsum/Emoji-Lipsum.utf16.txt
   6.629 ins/byte,    6.629 cycle/byte,    2.459 GB/s (0.7 %), 0.000092 b.misses/byte, 0.000000 c.mis/byte,    3.401 GHz,    4.793 ins/cycle
  26.515 ins/char,   26.515 cycle/char,    0.615 Gc/s (0.7 %), 0.000366 b.misses/char, 0.000000 c.mis/char,     4.00 byte/char
convert_utf16_to_utf32_with_dynamic_allocation+fallback, input size: 65542, iterations: 2000, dataset: unicode_lipsum/lipsum/Emoji-Lipsum.utf16.txt
   7.759 ins/byte,    7.759 cycle/byte,    2.216 GB/s (0.5 %), 0.000076 b.misses/byte, 0.000000 c.mis/byte,    3.400 GHz,    5.057 ins/cycle
  31.037 ins/char,   31.037 cycle/char,    0.554 Gc/s (0.5 %), 0.000305 b.misses/char, 0.000000 c.mis/char,     4.00 byte/char
convert_utf16_to_utf32_with_dynamic_allocation+haswell, input size: 65542, iterations: 2000, dataset: unicode_lipsum/lipsum/Emoji-Lipsum.utf16.txt
   6.916 ins/byte,    6.916 cycle/byte,    2.338 GB/s (2.4 %), 0.000076 b.misses/byte, 0.000000 c.mis/byte,    3.401 GHz,    4.755 ins/cycle
  27.665 ins/char,   27.665 cycle/char,    0.585 Gc/s (2.4 %), 0.000305 b.misses/char, 0.000000 c.mis/char,     4.00 byte/char
convert_utf16_to_utf32_with_dynamic_allocation+westmere, input size: 65542, iterations: 2000, dataset: unicode_lipsum/lipsum/Emoji-Lipsum.utf16.txt
   7.323 ins/byte,    7.323 cycle/byte,    2.205 GB/s (0.8 %), 0.000076 b.misses/byte, 0.000000 c.mis/byte,    3.400 GHz,    4.749 ins/cycle
  29.289 ins/char,   29.289 cycle/char,    0.551 Gc/s (0.8 %), 0.000305 b.misses/char, 0.000000 c.mis/char,     4.00 byte/char
dlemire@programming_station-for-dlemire:~/CVS/github/simdutf$ ./build/benchmarks/benchmark -F unicode_lipsum/lipsum/Emoji-Lipsum.utf16.txt -P convert_utf16_to_utf32
We define the number of bytes to be the number of *input* bytes.
We define a 'char' to be a code point (between 1 and 4 bytes).
===========================
Using ICU version 67.1
testcases: 1
input detected as UTF16 little-endian
current system detected as haswell
===========================
convert_utf16_to_utf32+fallback, input size: 65542, iterations: 2000, dataset: unicode_lipsum/lipsum/Emoji-Lipsum.utf16.txt
   6.503 ins/byte,    6.503 cycle/byte,    2.580 GB/s (2.3 %), 0.000061 b.misses/byte, 0.000000 c.mis/byte,    3.401 GHz,    4.933 ins/cycle
  26.013 ins/char,   26.013 cycle/char,    0.645 Gc/s (2.3 %), 0.000244 b.misses/char, 0.000000 c.mis/char,     4.00 byte/char
convert_utf16_to_utf32+haswell, input size: 65542, iterations: 2000, dataset: unicode_lipsum/lipsum/Emoji-Lipsum.utf16.txt
   6.598 ins/byte,    6.598 cycle/byte,    2.463 GB/s (2.6 %), 0.000061 b.misses/byte, 0.000000 c.mis/byte,    3.402 GHz,    4.777 ins/cycle
  26.390 ins/char,   26.390 cycle/char,    0.616 Gc/s (2.6 %), 0.000244 b.misses/char, 0.000000 c.mis/char,     4.00 byte/char
convert_utf16_to_utf32+llvm, input size: 65542, iterations: 2000, dataset: unicode_lipsum/lipsum/Emoji-Lipsum.utf16.txt
   5.254 ins/byte,    5.254 cycle/byte,    3.386 GB/s (0.3 %), 0.000061 b.misses/byte, 0.000000 c.mis/byte,    3.404 GHz,    5.226 ins/cycle
  21.015 ins/char,   21.015 cycle/char,    0.847 Gc/s (0.3 %), 0.000244 b.misses/char, 0.000000 c.mis/char,     4.00 byte/char
convert_utf16_to_utf32+westmere, input size: 65542, iterations: 2000, dataset: unicode_lipsum/lipsum/Emoji-Lipsum.utf16.txt
   6.629 ins/byte,    6.629 cycle/byte,    2.461 GB/s (0.7 %), 0.000061 b.misses/byte, 0.000000 c.mis/byte,    3.401 GHz,    4.797 ins/cycle
  26.515 ins/char,   26.515 cycle/char,    0.615 Gc/s (0.7 %), 0.000244 b.misses/char, 0.000000 c.mis/char,     4.00 byte/char
convert_utf16_to_utf32_with_dynamic_allocation+fallback, input size: 65542, iterations: 2000, dataset: unicode_lipsum/lipsum/Emoji-Lipsum.utf16.txt
   7.759 ins/byte,    7.759 cycle/byte,    2.218 GB/s (0.5 %), 0.000076 b.misses/byte, 0.000000 c.mis/byte,    3.400 GHz,    5.061 ins/cycle
  31.037 ins/char,   31.037 cycle/char,    0.554 Gc/s (0.5 %), 0.000305 b.misses/char, 0.000000 c.mis/char,     4.00 byte/char
convert_utf16_to_utf32_with_dynamic_allocation+haswell, input size: 65542, iterations: 2000, dataset: unicode_lipsum/lipsum/Emoji-Lipsum.utf16.txt
   6.916 ins/byte,    6.916 cycle/byte,    2.336 GB/s (2.4 %), 0.000076 b.misses/byte, 0.000000 c.mis/byte,    3.431 GHz,    4.710 ins/cycle
  27.665 ins/char,   27.665 cycle/char,    0.584 Gc/s (2.4 %), 0.000305 b.misses/char, 0.000000 c.mis/char,     4.00 byte/char
convert_utf16_to_utf32_with_dynamic_allocation+westmere, input size: 65542, iterations: 2000, dataset: unicode_lipsum/lipsum/Emoji-Lipsum.utf16.txt
   7.323 ins/byte,    7.323 cycle/byte,    2.203 GB/s (1.1 %), 0.000092 b.misses/byte, 0.000000 c.mis/byte,    3.400 GHz,    4.745 ins/cycle
  29.289 ins/char,   29.289 cycle/char,    0.551 Gc/s (1.1 %), 0.000366 b.misses/char, 0.000000 c.mis/char,     4.00 byte/char
```

lemire commented:
This PR looks good and we will merge it.

lemire commented:
Sync with our main branch to catch this commit: https://github.com/simdutf/simdutf/commit/24dc56e4c09f607e606d0f69108c61e873acb718

lemire commented on benchmarks/src/benchmark.cpp:
```diff
@@ -593,6 +732,32 @@ void Benchmark::run_convert_valid_utf8_to_utf16_utf8lut(size_t iterations) {
+    volatile size_t sink{0};
+    auto proc = [data, size, &output_buffer, &sink]() {
+      std::unique_ptr<BaseBufferProcessor> processor(ProcessorSelector<dfUtf8, dfUtf32>::WithOptions<cmFull>::Create());
+      ConversionResult result = ConvertInMemory(*processor, data, size, reinterpret_cast<char*>(output_buffer.get()), size*4+32);
```
This 4*size+32 is probably fine, but you allocate 4*(size+24).

lemire replied:
Note that we definitively want to overallocate by at least 16 bytes.

lemire commented on benchmarks/src/benchmark.cpp:
```diff
@@ -567,6 +680,32 @@ void Benchmark::run_convert_utf8_to_utf16_utf8lut(size_t iterations) {
+    volatile size_t sink{0};
+    auto proc = [data, size, &output_buffer, &sink]() {
+      std::unique_ptr<BaseBufferProcessor> processor(ProcessorSelector<dfUtf8, dfUtf32>::WithOptions<cmValidate>::Create());
+      ConversionResult result = ConvertInMemory(*processor, data, size, reinterpret_cast<char*>(output_buffer.get()), size*4+32);
```
You report `4*size+32` which is fine, but you allocated `4*(size+24)`.

NicolasJiaxin commented:
@lemire like this?

lemire commented:
>  like this?

I think not. I think you allocate `char32_t` and then you pass `char`, so you have to multiply your allocated region by four (since sizeof(char32_t)/sizeof(char) is 4).


I could be wrong... but I think that what we have looks ugly. It is not consistent.

NicolasJiaxin commented:
So, like this?

lemire commented:
This is now fine, I expect!

lemire commented:
I will handle the merge.

NicolasJiaxin commented:
It looks like it does not know in which order to merge both files. I can look into this @lemire

lemire commented:
It is fine. I think I will get through it. 

You should hear from me later today.
